### PR TITLE
DBZ-3 Creates a docker image to be used with the decoderbufs PG plugin

### PR DESCRIPTION
**Pull request by hchiorean**
The docker image uses the base Postgres 9.6 image and [my local decoderbufs fork](https://github.com/hchiorean/decoderbufs). 

The difference between the main repository and my fork is atm.:
* removal of the PostGIS dependency
* change of the Proto Java package names and classes

Ideally this Docker image should use a decoderbufs fork from the main [DBZ repository](https://github.com/debezium) which we can then modify as we see fit.

To validate that the docker image actually works, I've updated my [DBZ-3 branch](https://github.com/hchiorean/debezium/tree/DBZ-3) and the code for the [Postgres connector](https://github.com/hchiorean/debezium/tree/DBZ-3/debezium-connector-postgres) to use the `proto` file from the plugin, generate the appropriate Java classes (via a Maven plugin) and test that the Protobuf deserialization actually works.

The next step in the connector implementation is to get the updated PG driver with the [replication logic](https://github.com/pgjdbc/pgjdbc/pull/550) and implement the connector logic. Hopefully this PR makes it in the next PG driver release.

### CODE REVIEW

**Comment by rhauch** [comment]
@hchiorean, do you want to update this to use the [Debezium Postgres Decoderbufs](https://github.com/debezium/postgres-decoderbufs) library?

File: postgres/9.6/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# The plugin requires protobuf-c at least 1.1 but the stable Debian repo has an older version, so we need to add the Debian testing repo which has the correct version
+RUN add-apt-repository "deb http://ftp.debian.org/debian testing main contrib" && apt-get update
+ 
+# Install the protobuf-c dependency for the plugin
+RUN apt-get install -y libprotobuf-c-dev=1.2.1-1+b1
+ 
+# Compile the plugin from sources and install it
+# TODO: the plugin should probably be forked in the DBZ repository
+RUN git clone https://github.com/hchiorean/decoderbufs
+RUN cd /decoderbufs && make && make install && cd /
```
Comment (rhauch) [inline postgres/9.6/Dockerfile]: Should we "release" the code by making a tag, so the this build can check out that tag? Also, because of how Docker filesystems are layered, we probably want to use a single `RUN` command that also cleans up any intermediary files created by the build _and_ removes the local git repository. By doing all these steps on a single line, the only files that go into the new layer will be those that exist at the end of the line. Thoughts?

Comment (hchiorean) [inline postgres/9.6/Dockerfile]: >By doing all these steps on a single line, the only files that go into the new layer will be those that exist at the end of the line

I don't really know what this means :smile: 

> Also, because of how Docker filesystems are layered, we probably want to use a single RUN command that also cleans up any intermediary files created by the build and removes the local git repository.

I know Docker images area layers, but I don't know the implication of a single versus multiple `RUN` commands. As far as cleaning the files, I guess cleaning (i.e. removing) the Git repo is a good idea. Not sure what else can be removed

**Comment by hchiorean** [comment]
@rhauch: in addition to the above comment, this PR needs an update to use the debezium/postgres-decoderbufs code. On that topic - yes - I think having a release tag or branch and using that is a good idea (not using the latest master)

I'll also give another try at keeping the PostGIS code in and trying to get the Docker image to install these damned dependencies so that the plugin code is as complete as possible

File: postgres/9.6/Dockerfile (hunk 1 above)
Comment (rhauch) [inline postgres/9.6/Dockerfile]: The Docker filesystem is immutable and is comprised of a number of layers (starting with the last immutable layer in the `FROM` image). When Dockerfiles are run, each step results in a new layer in the image. So, the one of the [best practices](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/) is to minimize the number of layers (and therefore size of the image) by minimizing the number commands.

And, each layer created by the Dockerfile contains the _net change_ of the Dockerfile command. Layers are never removed (they can be by other optimizing tools), so when one command clones a git repo, that layer will have all of the files in that repo. No matter what happens next, the image will contain that layer with the complete git repo.

But, since the layer is produced to hold the _net change_ of each `RUN` command, you can minimize the layer by using a single `RUN` command to do multiple things, and if the `RUN` command cleans up any intermediary but ultimately unnecessary files before it completes, those are left out of the layer.

Bottom line is that I'm suggesting replacing your two `RUN` commands with the following:

```
RUN git clone https://github.com/hchiorean/decoderbufs && \
    cd /decoderbufs && \
    make && \
    make install && \
    cd / && \
    rm -rf decoderbufs
```

The first few commands in `RUN` are the same, but after installing the library into the postgres installation and changing to the root directory, it also removes the directory that contains the git clone and any intermediary files created by `make` and `make install`. The _net change_ of this `RUN` command should thus be just the installed libraries in the postgres installation.

File: postgres/9.6/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM postgres:9.6
+
+MAINTAINER Debezium Community
+
+# Install the packages which will be required to get everything to compile
+RUN apt-get update && apt-get install -y build-essential pkg-config git postgresql-server-dev-9.6 software-properties-common
```
Comment (rhauch) [inline postgres/9.6/Dockerfile]: A [best practice](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/) is to place each command in `RUN` on a separate line. Another is to place each installed dependency on a separate line. Both make it much easier to read complex run statements.

File: postgres/9.6/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+# The plugin requires protobuf-c at least 1.1 but the stable Debian repo has an older version, so we need to add the Debian testing repo which has the correct version
+RUN add-apt-repository "deb http://ftp.debian.org/debian testing main contrib" && apt-get update
+ 
+# Install the protobuf-c dependency for the plugin
+RUN apt-get install -y libprotobuf-c-dev=1.2.1-1+b1
```
Comment (rhauch) [inline postgres/9.6/Dockerfile]: You should probably run `rm -rf /var/lib/apt/lists/*` to clean up the `apt-get` cache populated by the multiple `apt-get` calls. But because of layering, cleaning the cache is only affective for files added to the cache during that one `RUN` statement, so for this reason you actually should combine all three of these `RUN` statements into a single one. This way, the resulting layer has only the dependencies that are needed below.

Technically, you could combine all of five of the `RUN` commands in a single one, but there are a couple of reasons I'd suggest not doing this. First, the first three existing `RUN` commands install dependencies, while the last two build and install the decoderbufs library into postgres, so it's useful to keep this distinction. Second, if the layer resulting from installing the dependencies were minimized, and the layer resulting from building and installing decoderbufs is minimized, combining these into an "uber" layer is not more efficient. And third, keeping a separate layers means that, when you build locally, Docker can cache the layers and will only have to build them if something changes -- something that's not frequent for the decoderbufs layer and even less likely for the dependency layer.

**Comment by rhauch** [comment]
@hchiorean, how about you get this stuff working, and then we can label the [Debezium Postgres Decoderbufs](https://github.com/debezium/postgres-decoderbufs) library and update this Dockerfile to use that label when building? We can do that as a separate PR if you want. (Note that Docker Hub won't actually build this Postgres Docker image until I manually add it via Docker Hub.)

**Comment by hchiorean** [comment]
@rhauch: after some grief, I'm also able to install some geom-related packages so that the `geography` and `geometry` types from PostGIS are working with the plugin

However, after doing some more investigation there are several drawbacks IMO to supporting these:

1. the support only applies to the `geography` and `geometry` types from the PostGIS library. I.E. this library has to be installed in Postgres for these types to be available; I.E. these _are not default Postgres types_
2. installing the required `geom-dev` packages for Debian almost doubles the Docker image size because of additional dependencies  - from 4xx to 7xx MB)
3. being a separate dependency, PostGIS is really an additional "burden" for DB/sys admins to install in Postgres, beside the "already intrusive" decoder plugin.

So based on the above (especially 1 and 3) my suggestion is to remove the PostGIS support altogether from the plugin and simply stick to the default Postgres types (note that not all PG types are supported atm but I suppose this is another topic...)

thoughts ?

**Comment by rhauch** [comment]
> @rhauch: after some grief, I'm also able to install some geom-related packages so that the `geography` and `geometry` types from PostGIS are working with the plugin

Awesome.

> 1. the support only applies to the `geography` and `geometry` types from the PostGIS library. I.E. this library has to be installed in Postgres for these types to be available; I.E. these _are not default Postgres types_

Correct. I do think we want to support this in decoderbufs (for now) and to provide a Docker image with PostGIS, simply because we want the _connector_ to be able to understand those types if they are used within the PostgreSQL server. Of course, if the PostgreSQL server doesn't have PostGIS installed, then the connector would never see those extra types in the logical decoding events. 

> 2. installing the required `geom-dev` packages for Debian almost doubles the Docker image size because of additional dependencies  - from 4xx to 7xx MB)
> 3. being a separate dependency, PostGIS is really an additional "burden" for DB/sys admins to install in Postgres, beside the "already intrusive" decoder plugin.

Yeah, I'm not too worried about the Docker image size for now. At some point in the future I expect we'll want to modify the decoderbufs code to exclude PostGIS as a _build-time option_ (so maintain a single code base), and also provide a Docker image for PostgreSQL with the decoderbufs plugin but without PostGIS. That would reduce the Docker image size for those that really don't want it, and would also enable DB/sys admins to choose which form of the decoderbufs plugin we really want.

In summary, now that you've got PostGIS working, I would prefer to include it for now. If the connector doesn't know about data types that people are using (including GIS types), then people can't use Debezium. IMO that's a far bigger concern than our decoderbufs plugin being "too big".

>  (note that not all PG types are supported atm but I suppose this is another topic...)

Agreed. In fact, could you document in [DBZ-3](https://issues.jboss.org/browse/DBZ-126) the types that will be supported? At some point we'll want to create an issue to support the remaining types.

**Comment by hchiorean** [comment]
ok; in that case can you pls tag/branch the `postgres-decoderbufs` repo so that it can be referred from the Docker image.

**Comment by rhauch** [comment]
I can certainly tag now (perhaps "0.1.0"?) but that means that if you find any bugs or want to make any changes then we'll need to "release" again. Alternatively, we could change this PR to use `master` of `postgres-decoderbufs`, and when we're getting closer to releasing the connector then tag `postgres-decoderbufs` and update this image.

I don't see one being better than the other, so I'll leave it up to you.

**Comment by hchiorean** [comment]
I don't see myself making any changes atm. to `postgres-decoderbufs` so perhaps tagging with `0.1.0` is the best option and using that in the docker image.

This docker image will likely _only_ be used by our tests, so it shouldn't be a problem IMO updating it. Actually it does raise the question of whether we want the docker image here (in this repo) or not: since it's not trivial to build (and time consuming) IMO it should definitively be in _some_ DBZ repo, published & built on DockerHub - as opposed to part of the DBZ PG connector's test resources.

That being said, I can see this as a test image not belonging in this repo and be subject to this repo's release cycle. So maybe we need another DBZ repo with test images which is more "liberal" release wise and won't necessary follow the lifecycle of the DBZ connector.

**Comment by rhauch** [comment]
Okay, I've tagged the `postgres-decoderbufs` with [`v0.1.0`](https://github.com/debezium/postgres-decoderbufs/tree/v0.1.0).

Even though we're going to use this primarily in our tests, I'm sure we'll also use it in future tutorials and examples, so I think it's fine to have this image defined here.

And the images in this repo are updated whenever necessary, not only when the Debezium code is released. For example, we can update our Java base image to use a newer patch release, and all of the other images will be rebuilt and updated (with whatever labels those images have). Or, we might change this image to use a newer patch release of PostgreSQL. Then, when somebody attempts to use an image they have locally, their local Docker will periodically check Docker Hub, notice the image has changed, and download and use the newer image.

**Comment by hchiorean** [comment]
@rhauch: I've amended the commit and updated the docker file  with the suggestions from above, plus PostGIS support and using the `v0.1.0` tag

**Comment by rhauch** [comment]
@hchiorean, this is excellent. Can you change `build-all` to call `build-postgres`? It's easier if it actually does build everything, and the extra time to run is negligible (especially after the first time).

**Comment by hchiorean** [comment]
amended the commit and added `build-postgres` to `build-all`

**Comment by rhauch** [comment]
@hchiorean thanks. I forgot that Travis-CI uses `build-all`, so that's a another reason for the change.

**Comment by hchiorean** [comment]
there's a `build-all.sh: line 18: ./build-postgres.sh: Permission denied` when trying to run the .sh script for Postgres; not sure why that is

**Comment by hchiorean** [comment]
@rhauch the Travis build did not run the Postgres script...

**Comment by rhauch** [comment]
It has the wrong permissions. I'll fix it.
